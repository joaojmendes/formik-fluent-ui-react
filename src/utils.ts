import { FieldProps, getIn } from 'formik'

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>

export function getErrorMessage<T = any>({ field, form }: FieldProps<T>) {
  const error = getIn(form.errors, field.name)
  const touched = getIn(form.touched, field.name)

  return touched ? error : undefined
}

export function invokeAll(...callbacks: any[]) {
  return () => {
    for (const callback of callbacks) {
      if (callback && typeof callback === 'function') {
        callback()
      }
    }
  }
}

export function createFakeEvent({ name }: { name: string }) {
  return { target: { name } }
}
