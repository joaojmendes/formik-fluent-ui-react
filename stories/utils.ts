import { action } from '@storybook/addon-actions'
import { FormikActions } from 'formik'

export function handleSubmit<T>(
  values: T,
  { setSubmitting }: FormikActions<T>
) {
  action('submit:start')(values)
  setTimeout(() => {
    setSubmitting(false)
    action('submit:done')()
  }, 2000)
}
