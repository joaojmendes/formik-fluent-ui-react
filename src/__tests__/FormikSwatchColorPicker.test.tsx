// tslint:disable:jsx-no-lambda

import { Field, FieldProps, Form, Formik } from 'formik'
import { IColorCellProps, SwatchColorPicker } from '@fluentui/react'
import * as React from 'react'
import renderer from 'react-test-renderer'
import {
  FormikSwatchColorPicker,
  mapFieldToSwatchColorPicker,
} from '../FormikSwatchColorPicker'
import { noop, serialize } from './utils'

const colors = [
  { id: 'red', label: 'red', color: '#ff0000' },
  { id: 'green', label: 'green', color: '#00ff00' },
  { id: 'blue', label: 'blue', color: '#0000ff' },
]

class Values {
  public color: IColorCellProps | string = colors[0]
}
function createFieldProps(
  value: IColorCellProps | string = colors[0]
): FieldProps<IColorCellProps | string> {
  return {
    field: {
      value,
      onChange: jest.fn(),
      onBlur: jest.fn(),
      name: 'color',
    },
    form: { setFieldValue: jest.fn(), handleBlur: jest.fn(() => jest.fn()) },
  } as any
}

test('<FormikSwatchColorPicker /> renders correctly as a field component', () => {
  const component = renderer.create(
    <Formik initialValues={new Values()} onSubmit={noop}>
      <Form>
        <Field
          name="color"
          render={(fieldProps: FieldProps<Values>) => (
            <FormikSwatchColorPicker
              {...fieldProps}
              columnCount={3}
              colorCells={[
                { id: '#ff0000', label: 'red', color: '#ff0000' },
                { id: '#00ff00', label: 'green', color: '#00ff00' },
                { id: '#0000ff', label: 'blue', color: '#0000ff' },
              ]}
            />
          )}
        />
      </Form>
    </Formik>
  )

  expect(component.toJSON()).toMatchSnapshot()
})

test('<FormikSwatchColorPicker /> renders a Fabric <SwatchColorPicker />', () => {
  const fieldProps = createFieldProps()

  const formikSwatchColorPicker = renderer.create(
    <FormikSwatchColorPicker
      {...fieldProps}
      colorCells={colors}
      columnCount={3}
    />
  )
  const fabricSwatchColorPicker = renderer.create(
    <SwatchColorPicker
      {...mapFieldToSwatchColorPicker(fieldProps)}
      colorCells={colors}
      columnCount={3}
    />
  )
  expect(serialize(formikSwatchColorPicker)).toBe(
    serialize(fabricSwatchColorPicker)
  )
})

test('mapFieldToSwatchColorPicker() maps FieldProps to ISwatchColorPickerProps', () => {
  const { field, form, meta } = createFieldProps()
  const props = mapFieldToSwatchColorPicker({ form, field, meta })

  expect(props.selectedId).toBe((field.value as { id: string, label: string, color: string }).id)

  props.onColorChanged!(colors[1].id, colors[1].color)

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, {
    id: colors[1].id,
    color: colors[1].color,
  })

  // onColorChanged also calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})

test('mapFieldToSwatchColorPicker() also supports string values', () => {
  const { field, form, meta } = createFieldProps(colors[0].color)
  const props = mapFieldToSwatchColorPicker({ form, field, meta })

  expect(props.selectedId).toBe(field.value)

  props.onColorChanged!(colors[1].id, colors[1].color)

  expect(form.setFieldValue).toHaveBeenCalledTimes(1)
  expect(form.setFieldValue).toHaveBeenCalledWith(field.name, colors[1].color)

  // onColorChanged also calls onBlur
  expect(field.onBlur).toHaveBeenCalled()
})
