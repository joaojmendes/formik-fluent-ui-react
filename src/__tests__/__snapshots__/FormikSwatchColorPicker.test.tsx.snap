// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FormikSwatchColorPicker /> renders correctly as a field component 1`] = `
<form
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="ms-FocusZone ms-swatchColorPickerBodyContainer focusedContainer-49"
    data-focuszone-id="FocusZone3"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDownCapture={[Function]}
    role="presentation"
  >
    <table
      className="root-50"
      id="id__2"
      onBlur={[Function]}
      role="grid"
    >
      <tbody>
        <tr
          role="row"
        >
          <td
            className="tableCell-48"
            role="presentation"
          >
            <button
              aria-label="red"
              aria-selected={false}
              className="ms-Button ms-Button--action ms-Button--command colorCell-53"
              data-index={0}
              data-is-focusable={true}
              id="swatchColorPicker1-#ff0000-0"
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onMouseUp={[Function]}
              role="gridcell"
              title="red"
              type="button"
            >
              <div
                className="ms-Button-flexContainer flexContainer-54"
              >
                <svg
                  className="svg-52"
                  fill="#ff0000"
                  viewBox="0 0 20 20"
                >
                  <circle
                    cx="50%"
                    cy="50%"
                    r="50%"
                  />
                </svg>
              </div>
            </button>
          </td>
          <td
            className="tableCell-48"
            role="presentation"
          >
            <button
              aria-label="green"
              aria-selected={false}
              className="ms-Button ms-Button--action ms-Button--command colorCell-53"
              data-index={1}
              data-is-focusable={true}
              id="swatchColorPicker1-#00ff00-1"
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onMouseUp={[Function]}
              role="gridcell"
              title="green"
              type="button"
            >
              <div
                className="ms-Button-flexContainer flexContainer-54"
              >
                <svg
                  className="svg-52"
                  fill="#00ff00"
                  viewBox="0 0 20 20"
                >
                  <circle
                    cx="50%"
                    cy="50%"
                    r="50%"
                  />
                </svg>
              </div>
            </button>
          </td>
          <td
            className="tableCell-48"
            role="presentation"
          >
            <button
              aria-label="blue"
              aria-selected={false}
              className="ms-Button ms-Button--action ms-Button--command colorCell-53"
              data-index={2}
              data-is-focusable={true}
              id="swatchColorPicker1-#0000ff-2"
              onClick={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseMove={[Function]}
              onMouseUp={[Function]}
              role="gridcell"
              title="blue"
              type="button"
            >
              <div
                className="ms-Button-flexContainer flexContainer-54"
              >
                <svg
                  className="svg-52"
                  fill="#0000ff"
                  viewBox="0 0 20 20"
                >
                  <circle
                    cx="50%"
                    cy="50%"
                    r="50%"
                  />
                </svg>
              </div>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</form>
`;
